### ISSUE TITLE
Generate Confusion Matrix of YOLOv8 Model

### ISSUE THREAD

**Issue by eddmar1993** [eddmar1993/YOLOv8#1]
I am a newbie in this area.  My apology. 
I am developing a YOLOv8 model for posture detection in pigs. I already have a trained model named "posture.pt".  The registered classes in data.yaml are Lateral Lying, Sitting, Standing, and Sternal Lying. I want to evaluate the model using the validation dataset found in "Z:\YOLO Training\Posture-3\valid\images" and the labels are found in "Z:\YOLO Training\Posture-3\valid\labels". To evaluate, **I want to get the Accuracy, Precision, Recall, and F1 Score for every posture class**. I am using Visual Studio Code to run the codes. The data should be saved in CSV format. The codes below are generated by an AI. However, I could not find the necessary utilities (utils/datasets.py, models/experimental.py, utils/general.py, and utils/torch_utils.py) to run these codes. Or is there any simple codes to generate the parameters abovementioned? I also included a screenshot of my VS Code project.
I am seeking your warm help. Thank you very much.

import torch
from torch.utils.data import DataLoader
from torchvision import transforms
from tqdm import tqdm
import numpy as np
import pandas as pd
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from utils.datasets import LoadImagesAndLabels
from models.experimental import attempt_load
from utils.general import non_max_suppression, scale_coords
from utils.torch_utils import select_device

# Load the YOLOv8 model
weights = 'path/to/posture.pt'
device = select_device('')
model = attempt_load(weights, map_location=device)
model.eval()

# Define the data directories
image_dir = 'Z:/YOLO Training/Posture-3/valid/images'
label_dir = 'Z:/YOLO Training/Posture-3/valid/labels'

# Load data
dataset = LoadImagesAndLabels(image_dir, label_dir, img_size=640)

# Define transformations
transform = transforms.Compose([
    transforms.ToTensor(),
])

# DataLoader
dataloader = DataLoader(dataset,
                        batch_size=1,
                        shuffle=False,
                        collate_fn=dataset.collate_fn)

# Evaluation loop
outputs = []
targets = []

for img, target, path, _ in tqdm(dataloader, desc="Evaluating"):
    img = torch.stack(img).to(device)
    _, _, h, w = img.shape
    img /= 255.0  # 0 - 255 to 0.0 - 1.0

    # Inference
    with torch.no_grad():
        detections = model(img)

    # Post-process detections
    detections = non_max_suppression(detections, 0.3, 0.5, agnostic=True)
    if detections[0] is not None:
        detections = detections[0]
        detections = scale_coords(img.shape[2:], detections[0][:, :4], (h, w)).round()
        outputs.append(detections.cpu().numpy())
    else:
        outputs.append([])

    # Get target labels
    targets.append(target)

# Convert outputs and targets to numpy arrays
outputs = np.array(outputs)
targets = np.array(targets)

# Define classes
classes = ['Lateral Lying', 'Sitting', 'Standing', 'Sternal Lying']

# Compute evaluation metrics
metrics = {'Class': [], 'Accuracy': [], 'Precision': [], 'Recall': [], 'F1 Score': []}
for i, class_name in enumerate(classes):
    y_true = np.array(targets == i, dtype=int).flatten()
    y_pred = np.array([1 if i in output[:, 5] else 0 for output in outputs], dtype=int)
    accuracy = accuracy_score(y_true, y_pred)
    precision = precision_score(y_true, y_pred)
    recall = recall_score(y_true, y_pred)
    f1 = f1_score(y_true, y_pred)

    metrics['Class'].append(class_name)
    metrics['Accuracy'].append(accuracy)
    metrics['Precision'].append(precision)
    metrics['Recall'].append(recall)
    metrics['F1 Score'].append(f1)

# Save metrics to CSV
metrics_df = pd.DataFrame(metrics)
metrics_df.to_csv('posture_evaluation_metrics.csv', index=False)

print("Evaluation completed. Metrics saved to posture_evaluation_metrics.csv")

![image](https://github.com/eddmar1993/YOLOv8/assets/148829286/e95c390b-771e-4956-be95-bb6be0d0bbb0)
